### ISSUE TITLE
Design Idea: Machine Learning (ML)

### ISSUE THREAD

**Issue by mokun** [mars-sim/mars-sim#183]
How likely can we incorporate machine learning (ML) into mars-sim ?

In neural network, a brain is no different than a 2-D decision matrix with weights that signify the strength between interconnections

How do we get start to create a brain for each settler?

We can create some initial training environment prior to sending the settlers to Mars. 

Running the settlers through the training environment is similar to run their matrix brain through some predefined training sets.

If they get a reasonably high score through these training, they are ready for the mission.

How should we construct the training sets ?

The next thing is to consider saving the settler's brain matrix in JSON in each simulation.

Running them in mars-sim will still adjust the weights in the brain matrix since the person should "grow" in the simulation.

This way, the saved settler may be reused on a different mission scenario. The player may pick and choice whoever he wants to keep. 

Is [DL4J](https://deeplearning4j.org/) a good library for us to adopt or pick a simpler library for experimentation ?

**Comment by Urwumpe** [comment]
Well, if it is optional, it could be interesting for small crews. Otherwise, it might be take too many CPU cycles. Also, abstracting the environment to something a small neural network can handle might be hard.

**Comment by mokun** [comment]
I just came across this [article](https://singularityhub.com/2019/02/19/giving-neural-nets-an-innate-brain-like-structure-could-bolster-deep-learning/#sm.0001g99l9x274czgw3g1syvnck0kn) talking about how the current AI research should borrow more from neuro-science if we want a true breakthrough in making our machine smart. 

Somewhere else I read that we may enter another AI winter or hit a AI bubble soon if ML or Deep Learning turns out they can't deliver what they promise.

**Comment by Urwumpe** [comment]
> Somewhere else I read that we may enter another AI winter or hit a AI bubble soon if ML or Deep Learning turns out they can't deliver what they promise.

Well, maybe even both at the same time. I some cases, it does already deliver impressive things. In others it fails despite all optimism. Just look at Facebooks algorithms there, which are hardly smart enough to work better than natural stupidity.

**Comment by mokun** [comment]
This article shared briefly the idea of [Neural MMO](https://blog.openai.com/neural-mmo/)

**Comment by mokun** [comment]
@Urwumpe , @shirishag75 , 

I came across this [video interview](https://www.closertotruth.com/series/what-things-are-conscious#video-49254) that discussed what we perceive to be conscious and non-conscious.

It's interesting that Scott pointed out that sometimes things give the appearance of being conscious when it's not.

So going back to mars-sim, my question is,

Q1: Self-Awareness : What are the top 3 things that do we want the settlers be aware of in the simulation ?

@Urwumpe you mentioned about the chat box in which the settlers gives random comments / complains, like in the PC game `Spacebase DF-9`.

Q2: ML Library : Which *light-weight* `Deep Learning` (DL)/ ML library in java can we employ for simulating a settler's mind ?

I know [DL4J](https://deeplearning4j.org/) being a very comprehensive library and easily added hundred of MB into mars-sim binary. So it's totally out of question that we can make use of it.

Q2: Can we come up with a few class prototyping how we would use JSON to store the mind of a settler using DL/ML?

**Comment by Urwumpe** [comment]
Q1: I would say social interaction is one important factor. Another is the awareness of being on Mars in contrast to Earth. We have a minimal health simulation, that could also be interesting as source for self-awareness - how does the settler feel?

Q2: I would say none, but again, that is less about not being able to create a model, than about the CPU/GGPU demands of it. We couldn't even simulate the mind of a single ant in real time with it. And we usually want to be faster than real time and want many more settlers than one. It is a topic for a far future version with different architecture and available hardware (PaaS?). And for the commander settler, the natural intelligence of the player is better, we really only need it for all non-human actors. 

For the current situation, I would say classic artificial non-intelligence is better, like Fuzzy Logic and Expert Systems. For example, we could allow settlers to create their own rules in their own knowledge base and then test them against the simulated reality. Something like that is fairly CPU friendly and we could also hardwire more necessary basic knowledge already to not let the settlers die of excessive stupidity. 

Classic expert systems are not emotional, but we want that so it might make sense to extend the concept with associating feelings with rules, like "doing this makes me happy" or "doing this makes me less afraid."

What might make sense, but I am not sure about the utility, would be describing personality and emotions with some kind of neutral network. As long as it is really small (<500 neurons) and we could live with some weird learning effects. 

Q3: Sure, but as said in Q2, I don't recommend it. Its way beyond our available hardware resources.

**Comment by mokun** [comment]
> classic artificial non-intelligence is better, like Fuzzy Logic and Expert Systems

Agree. We can definitely adopt a **hybrid** approach.

So we will keep the *Experience Points* and the *Skill Level* in `Skill`.

https://github.com/mars-sim/mars-sim/blob/9afdcbe5b7268c274e14099fb82ac993b93360fe/mars-sim-core/src/main/java/org/mars_sim/msp/core/person/ai/Skill.java#L55-L59

On top of that, say, we design a simple *Mind Matrix* based on neural network approach that can output the probability of getting that task done.

In my test of a 4x4 matrix, after feeding it with 40 to 60 training input vectors,  the output becomes stable enough. Say it's approaching between 0.9994 and 1.0054. 

So I'll take a few outputs and measure the deviation between them. 

Then we adjust the deviation with the experience points (and skill level) and translate the result into a probability score of getting a particular task done.

**Comment by mokun** [comment]
> but I am not sure about the utility, would be describing personality and emotions with some kind of neutral network. 

We can use the neural network approach to essentially **mimic** some sort of *learning processes* of the brain of a person performing tasks. 

So we are NOT here to create a *perfect mind* that can ace it with 100% success rate and no possibility of creating an accident/malfunction. 

Say, initially when we create a person, we train him with 500 iterations of `Botany` task (with some minor variations) to somewhat settle the weights inside the mind matrix. 

The inputs to a neural network can include environmental stimuli. 

So upon arrival, each time right after a person has performed `Botany` in a settlement on Mars, we would also feed him a few `Botany` task training vectors loaded with new stimuli based on the environment, the amount of sleep, relationship factors, etc. 

These become the feedback to his mind matrix. 

We could even create some training sets that mimic some sort of *diseases* or *traumatic events* that can *upset* or *distort* the weights inside the mind matrix, thus causing that person to lose his skill set or ability of getting something done.

**Comment by mokun** [comment]
Right now, I can only think of utilizing ML for improving Skill modeling.

Initially when we create a person, we train him with 500 iterations of `Botany` task (with some minor variations) to somewhat settle the weights inside the mind matrix. 

The inputs to a neural network can include environmental stimuli. 

So upon arrival on Mars, after a person has performed `Botany` in a settlement on Mars, we would also feed him a few `Botany` task *training vectors* loaded with new stimuli based on the environment, the amount of sleep, relationship factors, etc. 

These become the *feedback* to his mind matrix. 

In addition, we could even create some training sets that mimic lack of sleep, some sort of *diseases* he was inflicted with or *traumatic events* happened to him that can *upset* or *distort* the weights inside the mind matrix, thus causing him to temporarily lose a little bit of his skill ability of getting something done properly.

**Comment by mokun** [comment]
> would be describing personality and emotions with some kind of neutral network. 

May be in future, we can use DL to model multi-dimensional personality. 

Now each personality type is just a linear score. 

Emotion is already a 2D matrix in mars-sim. We can integrate it more into how it should affect relationship and other aspects of life.

May be we can figure out what we want ML to model to make all settlers to behave smarter.

**Comment by mokun** [comment]
@Urwumpe @shirishag75 @bevans2000 

This github [project](https://github.com/joonspk-research/generative_agents) are applying generative AI to create a small world where people have human-like conversation with one another.

They've written a [paper](https://arxiv.org/abs/2304.03442) to explain how they attempt to implement it.

**Comment by Urwumpe** [comment]
I think I have read about it on Mastodon. Don't ask me how to apply this to Mars-sim, but it would be nice for an experiment: Does this AI survive on Mars better than our own settlers?

**Comment by mokun** [comment]
> Does this AI survive on Mars better than our own settlers?

Great question. how would you like to rank the survivability of our current approach versus this generative AI approach and/or other algorithmic approach ?

I suppose what's impressive about this generative AI is its ability to parse language and its social implication. 

Being able to have conversation makes settlers very life-like and their interactions increasingly 'human' at say at least :)

**Comment by Urwumpe** [comment]
Yes, but it seems to be relying on ChatGPT, so no option for a longer operation if they restrict their service. 

Surviveability can we currently rate simply by number of settlers surviving per Mars year. Mars is currently as deadly as forced labour in a fascist regime...
